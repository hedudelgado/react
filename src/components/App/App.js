import React, { Component } from 'react';
import {
  array,
  object,
  func,
  bool,
  node,
} from 'prop-types';
import {
  PASSWORD_VIEW,
  PAYMENT_CARD_VIEW,
  USER_DETAILS_VIEW,
  ROOM_REQUIREMENTS_VIEW,
  EXTRAS_PREFERENCES_VIEW,
} from './appActions';
import { ComponentTogglerContainer } from '../ComponentToggler';
import UserProfile from '../UserProfile';
import Password from '../Password';
import UserProfileFormContainer from '../UserProfileForm';
import PasswordFormContainer from '../PasswordForm';
import { PaymentCardContainer } from '../PaymentCard';
import { RegularGuestsContainer } from '../RegularGuests';
import { NewsletterContainer } from '../Newsletter';
import Modal from '../../layout/Modal';
import PaymentCardFormContainer from '../PaymentCardForm';
import { DeleteCustomerContainer } from '../DeleteCustomer';
import RoomRequirements from '../RoomRequirements';
import RoomRequirementsFormContainer from '../RoomRequirementsForm';
import ExtrasPreferences from '../ExtrasPreferences';
import ExtrasPreferencesFormContainer from '../ExtrasPreferencesForm';
import BaseStyle from '../../layout/BaseStyle';
import MemorableWord from '../MemorableWord';
import MoveAccount from '../MoveAccount';
import {
  STAYER, SELF, managePermissionFlag, USER_LOGGED_IN_EVENT,
} from '../../config.json';
import ManagePermission from '../ManagePermission';

class App extends Component {
  constructor(props) {
    super(props);
    this.handleLoggedIn = this.handleLoggedIn.bind(this);
  }

  componentDidMount() {
    window.addEventListener('message', this.handleLoggedIn);
  }

  componentWillUnmount() {
    window.removeEventListener('message', this.handleLoggedIn);
  }

  async handleLoggedIn(event) {
    if (event.data.action === USER_LOGGED_IN_EVENT) {
      const { getSettings } = this.props;
      const { contactDetail: { email }, sessionId, business } = JSON.parse(event.data.profile);
      getSettings(email, sessionId, !!business);
    }
  }

  render() {
    const {
      countries,
      dictionary,
      profile,
      regions,
      newsletter,
      displayModal,
      handleModalClose,
      modalContent,
      userDataReceived,
      application: {
        showEmailPreferences,
      },
    } = this.props;

    const {
      business,
      business: {
        accessLevel = '',
        tetheredGuid = '',
      } = {},
    } = profile;

    const appReady = userDataReceived
      && (!showEmailPreferences || !!regions.length)
      && !!countries.length
      && (!showEmailPreferences || !!Object.keys(newsletter).length)
      && !!Object.keys(dictionary).length
      && !!Object.keys(profile).length;

    const showExtrasPreferences = !(accessLevel === SELF || accessLevel === STAYER);

    return (
      <div data-test="AccountSettings">
        {appReady
        && (
          <BaseStyle className="wb-typography app-container">
            <ComponentTogglerContainer
              viewName={USER_DETAILS_VIEW}
              view={UserProfile}
              edit={UserProfileFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PASSWORD_VIEW}
              view={Password}
              edit={PasswordFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PAYMENT_CARD_VIEW}
              view={PaymentCardContainer}
              edit={PaymentCardFormContainer}
            />
            {managePermissionFlag && (
              <ManagePermission dictionary={dictionary} />
            )}
            <RegularGuestsContainer />
            {showEmailPreferences && (
              <NewsletterContainer />
            )}
            <ComponentTogglerContainer
              viewName={ROOM_REQUIREMENTS_VIEW}
              view={RoomRequirements}
              edit={RoomRequirementsFormContainer}
            />
            {showExtrasPreferences && (
              <ComponentTogglerContainer
                viewName={EXTRAS_PREFERENCES_VIEW}
                view={ExtrasPreferences}
                edit={ExtrasPreferencesFormContainer}
              />
            )}
            {!!tetheredGuid && (
              <MemorableWord />
            )}
            {business && !tetheredGuid && (
              <MoveAccount />
            )}
            {!business && (
            <DeleteCustomerContainer />
            )}
            {displayModal && <Modal data-test="Modal" display content={modalContent} handleClose={handleModalClose} />}
          </BaseStyle>
        )
      }
      </div>
    );
  }
}

App.propTypes = {
  countries: array.isRequired,
  dictionary: object.isRequired,
  profile: object.isRequired,
  getSettings: func.isRequired,
  displayModal: bool.isRequired,
  handleModalClose: func.isRequired,
  modalContent: node,
  newsletter: object.isRequired,
  regions: array.isRequired,
  userDataReceived: bool.isRequired,
  application: object.isRequired,
};

export default App;
