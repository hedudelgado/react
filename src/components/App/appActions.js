import { HeaderInterface, InvalidSessionErrorCodeMap, InvalidSessionHelper } from 'mega-code';
import {
  getCountries,
  getNewsletterRegions,
  getNewsletterPreferences,
  getProfile,
  updateProfile,
  deleteCustomer,
} from '../../utils/api';

// actions
export const USER_LOGGED_IN_SUCCESSFULLY = 'USER_LOGGED_IN_SUCCESSFULLY';
export const COUNTRIES_RECEIVED = 'COUNTRIES_RECEIVED';
export const PROFILE_RECEIVED = 'PROFILE_RECEIVED';
export const REGIONS_RECEIVED = 'REGIONS_RECEIVED';
export const NEWSLETTER_RECEIVED = 'NEWSLETTER_RECEIVED';
export const SET_EDIT_VIEW = 'SET_EDIT_VIEW';
export const SET_NOTIFICATION_MESSAGE = 'SET_NOTIFICATION_MESSAGE';
export const PROFILE_UPDATED = 'PROFILE_UPDATED';
export const NEWSLETTER_UPDATED = 'NEWSLETTER_UPDATED';
export const RESET_NEWSLETTER_INFO = 'RESET_NEWSLETTER_INFO';
export const RESET_USER_INFO = 'RESET_USER_INFO';
// views
export const DEFAULT_VIEW = undefined;
export const USER_DETAILS_VIEW = 'USER_DETAILS_VIEW';
export const PASSWORD_VIEW = 'PASSWORD_VIEW';
export const PAYMENT_CARD_VIEW = 'PAYMENT_CARD_VIEW';
export const ADD_GUEST_VIEW = 'ADD_GUEST_VIEW';
export const REGULAR_GUESTS_VIEW = 'REGULAR_GUESTS_VIEW';
export const NEWSLETTER_VIEW = 'NEWSLETTER_VIEW';
export const ROOM_REQUIREMENTS_VIEW = 'ROOM_REQUIREMENTS_VIEW';
export const EXTRAS_PREFERENCES_VIEW = 'EXTRAS_PREFERENCES_VIEW';
export const MEMORABLE_WORD_VIEW = 'MEMORABLE_WORD_VIEW';
// errors
export const PROFILE_ERROR = 'PROFILE_ERROR';
export const PASSWORD_ERROR = 'PASSWORD_ERROR';
export const NEWSLETTER_ERROR = 'NEWSLETTER_ERROR';
export const PAYMENT_CARD_ERROR = 'PAYMENT_CARD_ERROR';
export const USER_DETAILS_ERROR = 'USER_DETAILS_ERROR';
export const REGULAR_GUESTS_ERROR = 'REGULAR_GUESTS_ERROR';
export const MEMORABLE_WORD_ERROR = 'MEMORABLE_WORD_ERROR';
export const ROOM_REQUIREMENTS_ERROR = 'ROOM_REQUIREMENTS_ERROR';
export const EXTRAS_PREFERENCES_ERROR = 'EXTRAS_PREFERENCES_ERROR';

const checkForInvalidSession = (error, config, dispatch) => {
  if (InvalidSessionHelper
    .isSessionExpired(error, config)) {
    dispatch({
      type: RESET_USER_INFO,
    });
    dispatch({
      type: RESET_NEWSLETTER_INFO,
    });
    HeaderInterface.userLogOut(HeaderInterface.LOG_OUT_REASONS.SESSION_EXPIRED);
  }
};

export const getSettings = (email, sessionId, business) => (dispatch, getState) => {
  getProfile(email, sessionId, business).then((profile) => {
    dispatch({
      type: PROFILE_RECEIVED,
      profile,
    });
  }).catch((error) => {
    checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
    dispatch({
      type: PROFILE_ERROR,
      error,
    });
  });
  getCountries()
    .then((countries) => {
      dispatch({
        type: COUNTRIES_RECEIVED,
        countries,
      });
    });
  if (getState().application.showEmailPreferences) {
    getNewsletterRegions()
      .then((regions) => {
        dispatch({
          type: REGIONS_RECEIVED,
          regions,
        });
      });

    getNewsletterPreferences(email)
      .then((newsletter) => {
        dispatch({
          type: NEWSLETTER_RECEIVED,
          newsletter,
        });
      })
      .catch(e => dispatch({
        type: NEWSLETTER_ERROR,
        error: e,
      }));
  }
};

export const updateUserProfile = payload => dispatch => updateProfile(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
  return Promise.reject(error);
});

export const deleteUserProfile = payload => dispatch => deleteCustomer(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.DELETE_PROFILE.invalidSession, dispatch);
  return Promise.reject(error);
});

export default {
  getSettings,
};
