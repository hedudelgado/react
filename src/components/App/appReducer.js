import {
  SET_NOTIFICATION_MESSAGE, SET_EDIT_VIEW, DEFAULT_VIEW,
} from './appActions';

const defaultState = {
  editViewName: DEFAULT_VIEW,
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case SET_EDIT_VIEW:
      return {
        ...state,
        editViewName: action.editViewName,
      };
    case SET_NOTIFICATION_MESSAGE:
      return {
        ...state,
        notificationMessage: action.notificationMessage,
      };
    default:
      return state;
  }
};
