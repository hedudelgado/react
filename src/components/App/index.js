export { default } from './App';

export { default as AppActions } from './appActions';

export { default as AppContainer } from './AppContainer';

export { default as modalReducer } from './modalReducer';
