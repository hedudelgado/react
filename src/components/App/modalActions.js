export const MODAL_OPEN = 'MODAL_OPEN';
export const MODAL_CLOSE = 'MODAL_CLOSE';

export const openModal = content => ({
  type: MODAL_OPEN,
  content,
});

export const closeModal = () => ({
  type: MODAL_CLOSE,
});
