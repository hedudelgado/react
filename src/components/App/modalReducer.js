import {
  MODAL_OPEN,
  MODAL_CLOSE,
} from './modalActions';

export default (state = {
  display: false,
  content: null,
}, action) => {
  const {
    type,
    content,
  } = action;

  switch (type) {
    case MODAL_OPEN:
      return {
        ...state,
        display: true,
        content,
      };
    case MODAL_CLOSE:
      return {
        ...state,
        display: false,
        content: null,
      };
    default:
      return state;
  }
};
