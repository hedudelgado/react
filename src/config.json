{
  "USER_LOGGED_IN_EVENT": "userLoggedIn",
  "STAYER": "STAYER",
  "SELF": "SELF",
  "BOOKER": "BOOKER",
  "SUPER": "SUPER",
  "managePermissionFlag": false,
  "managePermissionLink": "/gb/en/home.html",
  "apiErrorCodes": {
    "INCORRECT_PASSWORD": "012"
  },
  "defaultCountry": "GB",
  "marketingRegions": [
    1,
    2,
    3,
    4
  ],
  "mealOptions": [{
      "dictionaryKey": "mealoptions.pibreakfast",
      "code": "11"
    },
    {
      "dictionaryKey": "mealoptions.continentalbreakfast",
      "code": "12"
    },
    {
      "dictionaryKey": "mealoptions.mealdeal",
      "code": "17"
    },
    {
      "dictionaryKey": "mealoptions.nomeals",
      "code": "0"
    }
  ],
  "roomRequirements": {
    "adults": [
      1,
      2
    ],
    "children": [
      0,
      1,
      2
    ],
    "type": [{
        "dictionaryKey": "roomrequirements.type.double",
        "code": "DB"
      },
      {
        "dictionaryKey": "roomrequirements.type.single",
        "code": "SB"
      },
      {
        "dictionaryKey": "roomrequirements.type.twin",
        "code": "TWIN"
      },
      {
        "dictionaryKey": "roomrequirements.type.accessible",
        "code": "DIS"
      },
      {
        "dictionaryKey": "roomrequirements.type.family",
        "code": "FAM"
      }
    ],
    "defaults": {
      "adults": 1,
      "children": 0,
      "cotRequired": false,
      "type": "DB",
      "hotelBrand": "PI"
    }
  },
  "BUSINESS": "BUSINESS",
  "LEISURE": "HOME",
  "paymentCards": {
    "AC": {
      "name": "Mastercard Credit",
      "img": "/content/dam/global/booking/Mastercard.jpg"
    },
    "AM": {
      "name": "American Express",
      "img": "/content/dam/global/booking/AX.jpg"
    },
    "AT": {
      "name": "Business Account",
      "img": "/content/dam/global/booking/Business_Account.jpg"
    },
    "DI": {
      "name": "Diners Club",
      "img": "/content/dam/global/booking/dinersclub.jpg"
    },
    "EL": {
      "name": "Electron",
      "img": "/content/dam/global/booking/Electron_white_v.jpg"
    },
    "MA": {
      "name": "Maestro",
      "img": "/content/dam/global/booking/maestro.jpg"
    },
    "DL": {
      "name": "Visa Debit",
      "img": "/content/dam/global/booking/Visa_Debit.jpg"
    },
    "MD": {
      "name": "Mastercard Debit",
      "img": "/content/dam/global/booking/MD.jpg"
    },
    "VI": {
      "name": "Visa Credit",
      "img": "/content/dam/global/booking/VC.jpg"
    }
  }
}
