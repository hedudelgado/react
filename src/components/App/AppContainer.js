import { connect } from 'react-redux';
import { getSettings as getSettingsAction } from './appActions';
import { closeModal } from './modalActions';

import App from './App';

const mapStateToProps = ({
  modal: {
    display: displayModal,
    content: modalContent,
  },
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userProfile: {
    userDataReceived,
    profile,
  },
}) => ({
  profile,
  displayModal,
  modalContent,
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userDataReceived,
});

const mapDispatchToProps = dispatch => ({
  getSettings:
    (email, sessionId, business) => dispatch(getSettingsAction(email, sessionId, business)),
  handleModalClose: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
