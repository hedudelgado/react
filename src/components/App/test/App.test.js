import React from 'react';
import {
  shallow,
} from 'enzyme';
import App from '../App';
import profile from '../../../../__mocks__/profile.json';
import config from '../../../config.json';

const getSettings = jest.fn();
const dictionary = {
  test: 'test',
  'logged.out.error.heading': 'heading',
  'logged.out.error.body': 'body',
};
const displayModal = false;
const handleModalClose = jest.fn();
const countries = [{
  countryCode: 'GB',
  countryCodeISO: 'GB',
  countryLegend: 'United Kingdom (the)',
  passportRequired: false,
  dialingCode: '+44',
  flagImg: '/content/dam/global/flags/United-Kingdom.png',
}];

const application = {
  language: 'en',
};

const regions = [{
  id: '1',
  description: 'UK & Ireland',
  active: true,
}];
const newsletter = {
  regionSubscriptions: [],
};
const defaultProps = {
  getSettings,
  dictionary,
  newsletter,
  regions,
  countries,
  profile,
  application,
  displayModal,
  handleModalClose,
  userDataReceived: true,
};

describe('App shell', () => {
  describe('Layout', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<App {...defaultProps} />);
    });

    it('should render the Account Settings application', () => {
      expect(wrapper).toMatchSelector('[data-test="AccountSettings"]');
      expect(wrapper).toContainMatchingElement('.app-container');
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should not render the Modal component by default', () => {
      expect(wrapper).not.toContainMatchingElement('[data-test="Modal"]');
    });

    it('should render the Modal component', () => {
      defaultProps.displayModal = true;
      wrapper.setProps(defaultProps);
      expect(wrapper).toContainMatchingElement('[data-test="Modal"]');
    });

    it('should not show app container if the user data is not received', () => {
      wrapper.setProps({
        userDataReceived: false,
      });
      expect(wrapper.find('.app-container')).not.toExist();
    });

    it('should not show app container if showEmailPreferences is enabled but regions are not received', () => {
      wrapper.setProps({
        userDataReceived: true,
        application: {
          showEmailPreferences: true,
        },
        regions: [],
      });
      expect(wrapper.find('.app-container')).not.toExist();
    });

    it('should not show app container if showEmailPreferences is enabled but newsletter is not received', () => {
      wrapper.setProps({
        userDataReceived: true,
        application: {
          showEmailPreferences: true,
        },
        regions: [1],
        newsletter: {},
      });
      expect(wrapper.find('.app-container')).not.toExist();
    });

    it('should show delete profile option if user is not a business user', () => {
      wrapper.setProps({
        profile: {
          business: undefined,
        },
      });

      expect(wrapper.find('Connect(DeleteCustomer)')).toExist();
    });

    it('should not show delete profile button if user is a business user', () => {
      wrapper.setProps({
        profile: {
          business: {
            accessLevel: config.BOOKER,
          },
        },
      });

      expect(wrapper.find('Connect(DeleteCustomer)')).not.toExist();
    });

    it('should not show Newsletter if showEmailPreferences is not enabled', () => {
      wrapper.setProps({
        userDataReceived: true,
        application: {
          showEmailPreferences: false,
        },
        regions: [1],
        newsletter: { someProp: 'someVal' },
      });
      expect(wrapper.find('Connect(Newsletter)')).not.toExist();
    });

    it('should show Newsletter if showEmailPreferences is enabled', () => {
      wrapper.setProps({
        userDataReceived: true,
        application: {
          showEmailPreferences: true,
        },
        regions: [1],
        newsletter: { someProp: 'someVal' },
      });
      expect(wrapper.find('Connect(Newsletter)')).toExist();
    });

    it('should show Manage Permissions component when managePermissionFlag is true', () => {
      config.managePermissionFlag = true;
      wrapper = shallow(<App {...defaultProps} />);
      expect(wrapper).toContainExactlyOneMatchingElement('ManagePermission');
    });

    it('should not show Manage Permissions component when managePermissionFlag is false', () => {
      config.managePermissionFlag = false;
      wrapper = shallow(<App {...defaultProps} />);
      expect(wrapper).not.toContainMatchingElement('ManagePermission');
    });

    it('should show Memorable Word component when guid is defined', () => {
      const newProfile = {
        contactDetail: {
          email: 'test@mailinator.com',
        },
        business: {
          tetheredGuid: 'aaa',
        },
      };
      wrapper.setProps({
        profile: newProfile,
      });
      expect(wrapper).toContainExactlyOneMatchingElement('Connect(MemorableWord)');
    });

    it('should render the MoveAccount component for non tethered business accounts', () => {
      const newProfile = {
        contactDetail: {
          email: 'test@mailinator.com',
        },
        business: {
          somethingElse: 'aaa',
        },
      };
      wrapper.setProps({
        profile: newProfile,
      });
      expect(wrapper).toContainExactlyOneMatchingElement('Connect(MoveAccount)');
    });
  });

  describe('app behaviour', () => {
    let wrapper;
    beforeEach(() => {
      jest.clearAllMocks();
      wrapper = shallow(<App {...defaultProps} />);
    });

    it('should remove userLoggedIn event listener on unmount', () => {
      window.removeEventListener = jest.fn();
      wrapper.unmount();
      expect(window.removeEventListener).toHaveBeenCalled();
    });

    it('should call getSettings on userLoggedIn ', async () => {
      const map = {};
      window.addEventListener = jest.fn((event, cb) => {
        map[event] = cb;
      });
      shallow(<App
        {
          ...defaultProps
        }
      />);
      await map.message({
        data: {
          action: 'userLoggedIn',
          profile: JSON.stringify(profile),
        },
      }); expect(getSettings).toHaveBeenCalledWith('some_email@email.com', 'b7o4HMeLxWQn8YzG', false);
    });

    it('should not call getSettings when the message action is not userLoggedIn', async () => {
      const map = {};
      window.addEventListener = jest.fn((event, cb) => {
        map[event] = cb;
      });
      shallow(<App
        {
          ...defaultProps
        }
      />);
      await map.message({
        data: {
          action: 'otherAction',
          profile: JSON.stringify(profile),
        },
      }); expect(getSettings).not.toHaveBeenCalled();
    });
  });
});
