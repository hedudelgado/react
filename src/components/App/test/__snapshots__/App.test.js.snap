// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shell Layout should match snapshot 1`] = `
<div
  data-test="AccountSettings"
>
  <BaseStyle
    className="wb-typography app-container"
    component="div"
  >
    <Connect(ComponentToggler)
      edit={[Function]}
      view={[Function]}
      viewName="USER_DETAILS_VIEW"
    />
    <Connect(ComponentToggler)
      edit={[Function]}
      view={[Function]}
      viewName="PASSWORD_VIEW"
    />
    <Connect(ComponentToggler)
      edit={[Function]}
      view={[Function]}
      viewName="PAYMENT_CARD_VIEW"
    />
    <Connect(RegularGuests) />
    <Connect(ComponentToggler)
      edit={[Function]}
      view={[Function]}
      viewName="ROOM_REQUIREMENTS_VIEW"
    />
    <Connect(ComponentToggler)
      edit={[Function]}
      view={[Function]}
      viewName="EXTRAS_PREFERENCES_VIEW"
    />
    <Connect(DeleteCustomer) />
  </BaseStyle>
</div>
`;
